xds/internal/xdslbregistry/converter.go

```diff
@@ -0,0 +1,288 @@
+// list.
+type converter interface {
+	convertToServiceConfig([]byte, int) (json.RawMessage, bool, error)
+}
+
+const (
+	defaultRingHashMinSize = 1024
+	defaultRingHashMaxSize = 8 * 1024 * 1024 // 8M
+)
+
+func init() {
+	register("type.googleapis.com/envoy.extensions.load_balancing_policies.ring_hash.v3.RingHash", &ringHashConverter{})
+	register("type.googleapis.com/envoy.extensions.load_balancing_policies.round_robin.v3.RoundRobin", &roundRobinConverter{})
+	register("type.googleapis.com/envoy.extensions.load_balancing_policies.wrr_locality.v3.WrrLocality", &wrrLocalityConverter{})
+	register("type.googleapis.com/envoy.extensions.load_balancing_policies.client_side_weighted_round_robin.v3.ClientSideWeightedRoundRobin", &weightedRoundRobinConverter{})
+	register("type.googleapis.com/xds.type.v3.TypedStruct", &v3TypedStructConverter{})
+	register("type.googleapis.com/udpa.type.v1.TypedStruct", &v1TypedStructConverter{})
+}
+
+// register registers the converter to the map keyed on a proto type.
+func register(protoType string, c converter) {
+	m[strings.ToLower(protoType)] = c
+}
+
+// getConverter returns the converter registered with the given proto type. If
+// no converter is registered with the name, nil is returned.
+func getConverter(name string) converter {
+	if c, ok := m[strings.ToLower(name)]; ok {
+		return c
+	}
+	return nil
+}
+
+// ConvertToServiceConfig converts a proto Load Balancing Policy configuration
+// into a json string. Returns an error if:
+//   - no supported policy found
+//   - there is more than 16 layers of recursion in the configuration
+//   - a failure occurs when converting the policy
+func ConvertToServiceConfig(lbPolicy *v3clusterpb.LoadBalancingPolicy) (json.RawMessage, error) {
+	return convertToServiceConfig(lbPolicy, 0)
+}
+
+func convertToServiceConfig(lbPolicy *v3clusterpb.LoadBalancingPolicy, depth int) (json.RawMessage, error) {
+	// "Configurations that require more than 16 levels of recursion are
+	// considered invalid and should result in a NACK response." - A51
+	if depth > 15 {
+		return nil, fmt.Errorf("lb policy %v exceeds max depth supported: 16 layers", lbPolicy)
+	}
+
+	// "This function iterate over the list of policy messages in
+	// LoadBalancingPolicy, attempting to convert each one to gRPC form,
+	// stopping at the first supported policy." - A52
+	for _, policy := range lbPolicy.GetPolicies() {
+		policy.GetTypedExtensionConfig().GetTypedConfig().GetTypeUrl()
+		converter := getConverter(policy.GetTypedExtensionConfig().GetTypedConfig().GetTypeUrl())
+		// "Any entry not in the above list is unsupported and will be skipped."
+		// - A52
+		// This includes Least Request as well, since grpc-go does not support
+		// the Least Request Load Balancing Policy.
+		if converter == nil {
+			continue
+		}
+		json, cont, err := converter.convertToServiceConfig(policy.GetTypedExtensionConfig().GetTypedConfig().GetValue(), depth)
+		if cont {
+			continue
+		}
```

Comment:
Can we instead just use `json == nil` for this?  It's less explicit, but it would streamline the API and the intent seems very clear (plus it's all internal to the package).

Reply:
Ummmmm I don't think that makes sense here (switching bool to invariant of json, which happens to be able to be nil or not nil). Feels weird at every real error you need to return a populated empty byte slice?

Reply:
No.  It would only skip if `err == nil && json == nil`.

Reply:
Sounds good. Switched :).